### PR TITLE
Fix @return statements to use $this or static when relevant

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20290
| License       | MIT
| Doc PR        | n/a

see #20290

### CODE REVIEW

**Comment by ro0NL** [comment]
With #21047 i also spotted some cases where `@return FQCN|$this` is used. (FQCN from the current class file, ie. the same ;-))

I have no clear ref on this, but PhpStorm really doesnt seem to like it. Or at least not when used inconsistently so it seems :confused: 

Perhaps worth looking at.

File: src/Symfony/Bridge/Twig/NodeVisitor/Scope.php [hunk 1]
```diff
@@ -42,7 +42,7 @@ public function __construct(Scope $parent = null)
     /**
      * Opens a new child scope.
      *
-     * @return Scope
+     * @return $this
```
Comment (SpacePossum) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: this should be `@return self` right?
`this` - return exactly the same instance
`self` - return new instance of exactly same type
`static` - return new instance of same type or any child of type

**Comment by fabpot** [comment]
I would appreciate some careful review before merging as re-reading the PR several times allowed me to fix some issues I introduced. I know, this is quite painful to do, but let's try to do it well.

File: src/Symfony/Component/Config/Definition/Builder/NormalizationBuilder.php [hunk 2]
```diff
@@ -38,7 +38,7 @@ public function __construct(NodeDefinition $node)
      * @param string $key    The key to remap
      * @param string $plural The plural of the key in case of irregular plural
      *
-     * @return NormalizationBuilder
+     * @return $this
      */
     public function remap($key, $plural = null)
```
Comment (ro0NL) [inline src/Symfony/Component/Config/Definition/Builder/NormalizationBuilder.php]: `before()` (below) can be `$this|ExprBuilder`

File: src/Symfony/Component/Config/Loader/Loader.php [hunk 3]
```diff
@@ -57,7 +57,7 @@ public function import($resource, $type = null)
      * @param mixed       $resource A resource
      * @param string|null $type     The resource type or null if unknown
      *
-     * @return LoaderInterface A LoaderInterface instance
+     * @return $this|self
```
Comment (ro0NL) [inline src/Symfony/Component/Config/Loader/Loader.php]: Should be `$this|LoaderInterface`

File: src/Symfony/Component/Console/Helper/Helper.php [hunk 4]
```diff
@@ -35,7 +35,7 @@ public function setHelperSet(HelperSet $helperSet = null)
     /**
      * Gets the helper set associated with this helper.
      *
-     * @return HelperSet A HelperSet instance
+     * @return HelperSet
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Helper/Helper.php]: Should be `HelperSet|null`

File: src/Symfony/Component/EventDispatcher/Event.php [hunk 5]
```diff
@@ -81,7 +81,7 @@ public function setDispatcher(EventDispatcherInterface $dispatcher)
     /**
      * Returns the EventDispatcher that dispatches this Event.
      *
-     * @return EventDispatcherInterface
+     * @return EventDispatcher
```
Comment (ro0NL) [inline src/Symfony/Component/EventDispatcher/Event.php]: `EventDispatcherInterface` was right? The `@var` is wrong..

File: src/Symfony/Component/Finder/Expression/Expression.php [hunk 6]
```diff
@@ -27,7 +27,7 @@ class Expression implements ValueInterface
     /**
      * @param string $expr
      *
-     * @return Expression
+     * @return static
```
Comment (ro0NL) [inline src/Symfony/Component/Finder/Expression/Expression.php]: `self`

File: src/Symfony/Bridge/Twig/NodeVisitor/Scope.php [hunk 7]
```diff
@@ -42,7 +42,7 @@ public function __construct(Scope $parent = null)
     /**
      * Opens a new child scope.
      *
-     * @return Scope
+     * @return self
```
Comment (ro0NL) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: This seems to be inconsistent with a lot of changes below.. where it goes to `static` for `new self`.

File: src/Symfony/Component/Finder/Expression/Regex.php [hunk 8]
```diff
@@ -55,7 +55,7 @@ class Regex implements ValueInterface
     /**
      * @param string $expr
      *
-     * @return Regex
+     * @return static
```
Comment (ro0NL) [inline src/Symfony/Component/Finder/Expression/Regex.php]: `self`. Im stopping here for now.

File: src/Symfony/Bridge/Twig/NodeVisitor/Scope.php (hunk 7 above)
Comment (ogizanagi) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: If `new self` is used, it should be `self` anyway.

Comment (ro0NL) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: Agree.

File: src/Symfony/Bridge/Twig/NodeVisitor/Scope.php [hunk 9]
```diff
@@ -52,7 +52,7 @@ public function enter()
     /**
      * Closes current scope and returns parent one.
      *
-     * @return Scope|null
+     * @return self|null
```
Comment (ogizanagi) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: I'm not sure we should prefer using `self` over the class/interface name. IMHO, we should only distinguish `$this`, `static` and `classname` usages, but `self` does not bring anything more, so it isn't worth it.

File: src/Symfony/Component/Console/Helper/Table.php [hunk 10]
```diff
@@ -93,7 +93,7 @@ public static function setStyleDefinition($name, TableStyle $style)
      *
      * @param string $name The style name
      *
-     * @return TableStyle A TableStyle instance
+     * @return static
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Helper/Table.php]: This change looks wrong to me. `self::$styles` is an array of `TableStyle`.

File: src/Symfony/Bridge/Twig/NodeVisitor/Scope.php (hunk 9 above)
Comment (ro0NL) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: > I'm not sure we should prefer using self over the class/interface name.

I'd prefer it.. so while we're at it :+1: for doing this as well.

File: src/Symfony/Component/CssSelector/Exception/SyntaxErrorException.php [hunk 11]
```diff
@@ -27,7 +27,7 @@ class SyntaxErrorException extends ParseException
      * @param string $expectedValue
      * @param Token  $foundToken
      *
-     * @return SyntaxErrorException
+     * @return static
```
Comment (ogizanagi) [inline src/Symfony/Component/CssSelector/Exception/SyntaxErrorException.php]: Should be `self` (same below)

File: src/Symfony/Component/Form/FormBuilderInterface.php [hunk 12]
```diff
@@ -38,7 +38,7 @@ public function add($child, $type = null, array $options = array());
      * @param string|FormTypeInterface $type    The type of the form or null if name is a property
      * @param array                    $options The options
      *
-     * @return FormBuilderInterface The created builder
+     * @return $this
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/FormBuilderInterface.php]: This change looks wrong to me. It doesn't return the current instance, but create another one, for a child for instance.

File: src/Symfony/Component/Form/FormBuilderInterface.php [hunk 13]
```diff
@@ -47,7 +47,7 @@ public function create($name, $type = null, array $options = array());
      *
      * @param string $name The name of the child
      *
-     * @return FormBuilderInterface The builder for the child
+     * @return $this
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/FormBuilderInterface.php]: Same. It returns a child. Not the same instance.

File: src/Symfony/Component/Form/FormConfigBuilder.php [hunk 14]
```diff
@@ -715,7 +715,7 @@ public function setInheritData($inheritData)
      *
      * @param bool $inheritData Whether the form should inherit its parent's data
      *
-     * @return FormConfigBuilder The configuration object
+     * @return $this
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/FormConfigBuilder.php]: It doesn't return anything actually.

File: src/Symfony/Component/Form/FormConfigBuilder.php [hunk 15]
```diff
@@ -349,7 +349,7 @@ public function getInheritData()
     /**
      * Alias of {@link getInheritData()}.
      *
-     * @return FormConfigBuilder The configuration object
+     * @return $this
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/FormConfigBuilder.php]: Hmmm. It returns a boolean actually.

File: src/Symfony/Bridge/Twig/NodeVisitor/Scope.php (hunk 9 above)
Comment (ogizanagi) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: >Personally, i prefer it...

I don't 😆 

I just don't understand the motivations behind, and there are many places where this was not changed.

File: src/Symfony/Component/Intl/Collator/Collator.php [hunk 16]
```diff
@@ -86,7 +86,7 @@ public function __construct($locale)
      *
      * @param string $locale The locale code. The only currently supported locale is "en" (or null using the default locale, i.e. "en")
      *
-     * @return Collator
+     * @return static
```
Comment (ogizanagi) [inline src/Symfony/Component/Intl/Collator/Collator.php]: self?

File: src/Symfony/Component/Intl/DateFormatter/IntlDateFormatter.php [hunk 17]
```diff
@@ -171,7 +171,7 @@ public function __construct($locale, $datetype, $timetype, $timezone = null, $ca
      *                         One of the calendar constants.
      * @param string $pattern  Optional pattern to use when formatting
      *
-     * @return IntlDateFormatter
+     * @return static
```
Comment (ogizanagi) [inline src/Symfony/Component/Intl/DateFormatter/IntlDateFormatter.php]: self?

File: src/Symfony/Component/Intl/NumberFormatter/NumberFormatter.php [hunk 18]
```diff
@@ -302,7 +302,7 @@ public function __construct($locale = 'en', $style = null, $pattern = null)
      *                        NumberFormat::PATTERN_RULEBASED. It must conform to  the syntax
      *                        described in the ICU DecimalFormat or ICU RuleBasedNumberFormat documentation
      *
-     * @return NumberFormatter
+     * @return static
```
Comment (ogizanagi) [inline src/Symfony/Component/Intl/NumberFormatter/NumberFormatter.php]: self?

File: src/Symfony/Component/PropertyAccess/PropertyAccess.php [hunk 19]
```diff
@@ -21,7 +21,7 @@
     /**
      * Creates a property accessor with the default configuration.
      *
-     * @return PropertyAccessor The new property accessor
+     * @return static
```
Comment (ogizanagi) [inline src/Symfony/Component/PropertyAccess/PropertyAccess.php]: Looks wrong. (same for other changes in this class)

File: src/Symfony/Component/Security/Acl/Domain/ObjectIdentity.php [hunk 20]
```diff
@@ -52,7 +52,7 @@ public function __construct($identifier, $type)
      *
      * @param object $domainObject
      *
-     * @return ObjectIdentity
+     * @return static
```
Comment (ogizanagi) [inline src/Symfony/Component/Security/Acl/Domain/ObjectIdentity.php]: self?

File: src/Symfony/Component/Security/Acl/Domain/UserSecurityIdentity.php [hunk 21]
```diff
@@ -52,7 +52,7 @@ public function __construct($username, $class)
      *
      * @param UserInterface $user
      *
-     * @return UserSecurityIdentity
+     * @return static
```
Comment (ogizanagi) [inline src/Symfony/Component/Security/Acl/Domain/UserSecurityIdentity.php]: self?

File: src/Symfony/Component/Security/Acl/Domain/UserSecurityIdentity.php [hunk 22]
```diff
@@ -64,7 +64,7 @@ public static function fromAccount(UserInterface $user)
      *
      * @param TokenInterface $token
      *
-     * @return UserSecurityIdentity
+     * @return static
      */
     public static function fromToken(TokenInterface $token)
```
Comment (ogizanagi) [inline src/Symfony/Component/Security/Acl/Domain/UserSecurityIdentity.php]: self?

**Review by ogizanagi — COMMENTED** [review]
I'm done 😅 
Sorry, I've left a lot of comment related to what seems misleading usages of `static` to me (see https://github.com/symfony/symfony/pull/21054#discussion_r93865721), in order to keep track of it. If you don't agree, I'll remove them.

Anyway, what a colossal job you've achieved here!

File: src/Symfony/Bridge/Twig/NodeVisitor/Scope.php (hunk 9 above)
Comment (ro0NL) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: Avoid duplicating terms. But you're right, if not done consistently it's not really worth it anyway.

File: src/Symfony/Component/Finder/Shell/Command.php [hunk 23]
```diff
@@ -61,7 +61,7 @@ public function __toString()
      *
      * @param Command|null $parent Parent command
      *
-     * @return Command New Command instance
+     * @return static
```
Comment (wouterj) [inline src/Symfony/Component/Finder/Shell/Command.php]: self

File: src/Symfony/Component/Finder/Shell/Command.php [hunk 24]
```diff
@@ -157,7 +157,7 @@ public function cmd($esc)
      *
      * @param string $label The unique label
      *
-     * @return Command The current Command instance
+     * @return self
```
Comment (wouterj) [inline src/Symfony/Component/Finder/Shell/Command.php]: Should be `self|string`, according to `add()`.

File: src/Symfony/Component/Finder/Shell/Command.php [hunk 25]
```diff
@@ -194,7 +194,7 @@ public function get($label)
     /**
      * Returns parent command (if any).
      *
-     * @return Command Parent command
+     * @return self
```
Comment (wouterj) [inline src/Symfony/Component/Finder/Shell/Command.php]: `self|null`

File: src/Symfony/Component/Finder/Shell/Command.php [hunk 26]
```diff
@@ -178,7 +178,7 @@ public function ins($label)
      *
      * @param string $label
      *
-     * @return Command The labeled command
+     * @return self
```
Comment (wouterj) [inline src/Symfony/Component/Finder/Shell/Command.php]: `self|string`

File: src/Symfony/Component/Form/FormInterface.php [hunk 27]
```diff
@@ -36,7 +36,7 @@ public function setParent(FormInterface $parent = null);
     /**
      * Returns the parent form.
      *
-     * @return FormInterface|null The parent form or null if there is none
+     * @return $this|null The parent form or null if there is none
```
Comment (wouterj) [inline src/Symfony/Component/Form/FormInterface.php]: Should be `self|null` then

File: src/Symfony/Component/Form/FormInterface.php [hunk 28]
```diff
@@ -80,7 +80,7 @@ public function has($name);
      *
      * @param string $name The name of the child to remove
      *
-     * @return FormInterface The form instance
+     * @return self
```
Comment (wouterj) [inline src/Symfony/Component/Form/FormInterface.php]: It's actually `$this`.

File: src/Symfony/Component/Form/Guess/Guess.php [hunk 29]
```diff
@@ -70,7 +70,7 @@
      *
      * @param Guess[] $guesses An array of guesses
      *
-     * @return Guess|null The guess with the highest confidence
+     * @return static|null
```
Comment (wouterj) [inline src/Symfony/Component/Form/Guess/Guess.php]: Or `self|null`

File: src/Symfony/Component/HttpKernel/Profiler/Profile.php [hunk 30]
```diff
@@ -191,7 +191,7 @@ public function getStatusCode()
     /**
      * Finds children profilers.
      *
-     * @return Profile[] An array of Profile
+     * @return Profile[]
```
Comment (wouterj) [inline src/Symfony/Component/HttpKernel/Profiler/Profile.php]: `self[]` ?

**Review by wouterj — COMMENTED** [review]
Did just a quick review, these are the glitches I found.

File: src/Symfony/Component/Console/Helper/Table.php (hunk 10 above)
Comment (ro0NL) [inline src/Symfony/Component/Console/Helper/Table.php]: This returns `self::$styles[$name]` or throws if not found. It's correct.

File: src/Symfony/Component/Config/Definition/Builder/NormalizationBuilder.php (hunk 2 above)
Comment (fabpot) [inline src/Symfony/Component/Config/Definition/Builder/NormalizationBuilder.php]: fixed

File: src/Symfony/Component/Config/Loader/Loader.php (hunk 3 above)
Comment (fabpot) [inline src/Symfony/Component/Config/Loader/Loader.php]: fixed

File: src/Symfony/Component/Console/Helper/Helper.php (hunk 4 above)
Comment (fabpot) [inline src/Symfony/Component/Console/Helper/Helper.php]: fixed

File: src/Symfony/Component/EventDispatcher/Event.php (hunk 5 above)
Comment (fabpot) [inline src/Symfony/Component/EventDispatcher/Event.php]: fixed

File: src/Symfony/Component/Finder/Expression/Expression.php (hunk 6 above)
Comment (fabpot) [inline src/Symfony/Component/Finder/Expression/Expression.php]: No, the method is static

Comment (fabpot) [inline src/Symfony/Component/Finder/Expression/Expression.php]: fixed

File: src/Symfony/Component/Finder/Expression/Regex.php (hunk 8 above)
Comment (fabpot) [inline src/Symfony/Component/Finder/Expression/Regex.php]: fixed

File: src/Symfony/Bridge/Twig/NodeVisitor/Scope.php (hunk 9 above)
Comment (fabpot) [inline src/Symfony/Bridge/Twig/NodeVisitor/Scope.php]: I will keep `self` as in the code, that's also what we are using.

File: src/Symfony/Component/Console/Helper/Table.php (hunk 10 above)
Comment (fabpot) [inline src/Symfony/Component/Console/Helper/Table.php]: fixed

File: src/Symfony/Component/CssSelector/Exception/SyntaxErrorException.php (hunk 11 above)
Comment (fabpot) [inline src/Symfony/Component/CssSelector/Exception/SyntaxErrorException.php]: fixed

File: src/Symfony/Component/Form/FormBuilderInterface.php (hunk 12 above)
Comment (fabpot) [inline src/Symfony/Component/Form/FormBuilderInterface.php]: fixed

File: src/Symfony/Component/Form/FormBuilderInterface.php (hunk 13 above)
Comment (fabpot) [inline src/Symfony/Component/Form/FormBuilderInterface.php]: fixed

File: src/Symfony/Component/Form/FormConfigBuilder.php (hunk 15 above)
Comment (fabpot) [inline src/Symfony/Component/Form/FormConfigBuilder.php]: indeed :)

File: src/Symfony/Component/Form/FormConfigBuilder.php (hunk 14 above)
Comment (fabpot) [inline src/Symfony/Component/Form/FormConfigBuilder.php]: fixed

File: src/Symfony/Component/Form/FormInterface.php (hunk 27 above)
Comment (fabpot) [inline src/Symfony/Component/Form/FormInterface.php]: fixed

File: src/Symfony/Component/Form/FormInterface.php [hunk 31]
```diff
@@ -89,7 +89,7 @@ public function remove($name);
     /**
      * Returns all children in this group.
      *
-     * @return FormInterface[] An array of FormInterface instances
+     * @return FormInterface[]
```
Comment (fabpot) [inline src/Symfony/Component/Form/FormInterface.php]: fixed

File: src/Symfony/Component/Form/Guess/Guess.php (hunk 29 above)
Comment (fabpot) [inline src/Symfony/Component/Form/Guess/Guess.php]: fixed

File: src/Symfony/Component/HttpFoundation/AcceptHeader.php [hunk 32]
```diff
@@ -48,7 +48,7 @@ public function __construct(array $items)
      *
      * @param string $headerValue
      *
-     * @return AcceptHeader
+     * @return static
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/AcceptHeader.php]: fixed

File: src/Symfony/Component/HttpFoundation/AcceptHeaderItem.php [hunk 33]
```diff
@@ -57,7 +57,7 @@ public function __construct($value, array $attributes = array())
      *
      * @param string $itemValue
      *
-     * @return AcceptHeaderItem
+     * @return static
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/AcceptHeaderItem.php]: fixed

File: src/Symfony/Component/HttpFoundation/File/MimeType/ExtensionGuesser.php [hunk 34]
```diff
@@ -42,7 +42,7 @@ class ExtensionGuesser implements ExtensionGuesserInterface
     /**
      * Returns the singleton instance.
      *
-     * @return ExtensionGuesser
+     * @return static
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/File/MimeType/ExtensionGuesser.php]: fixed

File: src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeGuesser.php [hunk 35]
```diff
@@ -56,7 +56,7 @@ class MimeTypeGuesser implements MimeTypeGuesserInterface
     /**
      * Returns the singleton instance.
      *
-     * @return MimeTypeGuesser
+     * @return static
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeGuesser.php]: fixed

File: src/Symfony/Component/Intl/Collator/Collator.php (hunk 16 above)
Comment (fabpot) [inline src/Symfony/Component/Intl/Collator/Collator.php]: fixed

File: src/Symfony/Component/Intl/DateFormatter/IntlDateFormatter.php (hunk 17 above)
Comment (fabpot) [inline src/Symfony/Component/Intl/DateFormatter/IntlDateFormatter.php]: fixed

File: src/Symfony/Component/Intl/NumberFormatter/NumberFormatter.php (hunk 18 above)
Comment (fabpot) [inline src/Symfony/Component/Intl/NumberFormatter/NumberFormatter.php]: fixed

File: src/Symfony/Component/PropertyAccess/PropertyAccess.php (hunk 19 above)
Comment (fabpot) [inline src/Symfony/Component/PropertyAccess/PropertyAccess.php]: fixed

File: src/Symfony/Component/Security/Acl/Domain/ObjectIdentity.php (hunk 20 above)
Comment (fabpot) [inline src/Symfony/Component/Security/Acl/Domain/ObjectIdentity.php]: fixed

File: src/Symfony/Component/Security/Acl/Domain/UserSecurityIdentity.php (hunk 21 above)
Comment (fabpot) [inline src/Symfony/Component/Security/Acl/Domain/UserSecurityIdentity.php]: fixed

File: src/Symfony/Component/Security/Acl/Domain/UserSecurityIdentity.php (hunk 22 above)
Comment (fabpot) [inline src/Symfony/Component/Security/Acl/Domain/UserSecurityIdentity.php]: fixed

File: src/Symfony/Component/Finder/Shell/Command.php (hunk 23 above)
Comment (fabpot) [inline src/Symfony/Component/Finder/Shell/Command.php]: fixed

File: src/Symfony/Component/Finder/Shell/Command.php (hunk 24 above)
Comment (fabpot) [inline src/Symfony/Component/Finder/Shell/Command.php]: fixed

File: src/Symfony/Component/Finder/Shell/Command.php (hunk 26 above)
Comment (fabpot) [inline src/Symfony/Component/Finder/Shell/Command.php]: fixed

File: src/Symfony/Component/Finder/Shell/Command.php (hunk 25 above)
Comment (fabpot) [inline src/Symfony/Component/Finder/Shell/Command.php]: I don't think so as it throws an exception is parent is `null`.

File: src/Symfony/Component/Form/FormInterface.php (hunk 28 above)
Comment (fabpot) [inline src/Symfony/Component/Form/FormInterface.php]: fixed

File: src/Symfony/Component/HttpKernel/Profiler/Profile.php (hunk 30 above)
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Profiler/Profile.php]: fixed

**Comment by fabpot** [comment]
@wouterj @ogizanagi @ro0NL Thank you very much for your reviews. I think we're good now.

**Comment by ro0NL** [comment]
@fabpot are you planning to do all branches?

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 36]
```diff
@@ -422,7 +422,7 @@ public static function setFactory($callable)
      * @param array $files      The FILES parameters
      * @param array $server     The SERVER parameters
      *
-     * @return Request The duplicated request
+     * @return self
```
Comment (ogizanagi) [inline src/Symfony/Component/HttpFoundation/Request.php]: I think this one was missed ;)

**Pull request merged**